File: app/Models/Community.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Str;

class Community extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'name',
        'slug',
        'description',
        'logo',
        'website',
        'contact_email',
        'calendar_email',
        'destination_calendar_id',
        'google_account_id',
        'social_links',
        'timezone',
        'color',
        'is_public',
        'is_active',
    ];

    protected $casts = [
        'social_links' => 'array',
        'is_public' => 'boolean',
        'is_active' => 'boolean',
    ];

    protected static function boot()
    {
        parent::boot();

        static::creating(function ($community) {
            if (empty($community->slug)) {
                $community->slug = Str::slug($community->name);

                $originalSlug = $community->slug;
                $counter = 1;
                while (static::where('slug', $community->slug)->exists()) {
                    $community->slug = $originalSlug . '-' . $counter;
                    $counter++;
                }
            }
        });
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function events(): HasMany
    {
        return $this->hasMany(CommunityEvent::class);
    }

    public function callsForSpeakers(): HasMany
    {
        return $this->hasMany(CallForSpeakers::class);
    }

    public function speakers(): HasMany
    {
        return $this->hasMany(EventSpeaker::class, 'community_event_id');
    }

    public function googleAccount(): BelongsTo
    {
        return $this->belongsTo(GoogleAccount::class);
    }

    public function destinationCalendar(): BelongsTo
    {
        return $this->belongsTo(Calendar::class, 'destination_calendar_id');
    }

    public function calendar()
    {
        return $this->hasOneThrough(
            Calendar::class,
            GoogleAccount::class,
            'user_id',
            'google_account_id',
            'user_id',
            'id'
        )->where('calendars.name', 'like', '%' . $this->name . '%')
            ->orWhere('calendars.is_primary', true);
    }

    public function getEffectiveDestinationCalendar()
    {
        if ($this->destination_calendar_id) {
            return Calendar::find($this->destination_calendar_id);
        }

        if ($this->google_account_id) {
            return Calendar::where('google_account_id', $this->google_account_id)
                ->where('is_primary', true)
                ->first();
        }

        $primaryAccount = GoogleAccount::where('user_id', $this->user_id)
            ->where('is_primary', true)
            ->first();

        if (!$primaryAccount) {
            $primaryAccount = GoogleAccount::where('user_id', $this->user_id)
                ->where('is_active', true)
                ->first();
        }

        if ($primaryAccount) {
            return Calendar::where('google_account_id', $primaryAccount->id)
                ->where('is_primary', true)
                ->first();
        }

        return null;
    }

    public function getAvailableCalendars()
    {
        $accountIds = collect();

        if ($this->google_account_id) {
            $accountIds->push($this->google_account_id);
        }

        $userAccounts = GoogleAccount::where('user_id', $this->user_id)
            ->where('is_active', true)
            ->pluck('id');

        $accountIds = $accountIds->merge($userAccounts);

        return Calendar::whereIn('google_account_id', $accountIds->unique())
            ->with('googleAccount:id,name,email,is_primary')
            ->get();
    }

    public function getEffectiveEventCalendar()
    {
        if ($this->destination_calendar_id) {
            $calendar = Calendar::find($this->destination_calendar_id);
            if ($calendar && $calendar->googleAccount->is_active) {
                return $calendar;
            }
        }

        if ($this->google_account_id) {
            $calendar = Calendar::where('google_account_id', $this->google_account_id)
                ->where('is_primary', true)
                ->first();

            if ($calendar) {
                return $calendar;
            }

            $calendar = Calendar::where('google_account_id', $this->google_account_id)
                ->first();

            if ($calendar) {
                return $calendar;
            }
        }

        $primaryAccount = GoogleAccount::where('user_id', $this->user_id)
            ->where('is_primary', true)
            ->where('is_active', true)
            ->first();

        if ($primaryAccount) {
            return Calendar::where('google_account_id', $primaryAccount->id)
                ->where('is_primary', true)
                ->first();
        }

        return null;
    }

    public function getAvailabilityCalendars()
    {
        if (empty($this->availability_calendars)) {
            $userAccounts = GoogleAccount::where('user_id', $this->user_id)
                ->where('is_active', true)
                ->pluck('id');

            return Calendar::whereIn('google_account_id', $userAccounts)
                ->where('is_visible', true)
                ->get();
        }

        return Calendar::whereIn('id', $this->availability_calendars)
            ->with('googleAccount')
            ->get();
    }

    public function hasCalendarIntegration()
    {
        return !empty($this->calendar_email) || !empty($this->google_account_id);
    }

    public function getCalendarAccount()
    {
        if ($this->google_account_id) {
            return GoogleAccount::find($this->google_account_id);
        }

        if ($this->calendar_email) {
            return GoogleAccount::where('user_id', $this->user_id)
                ->where('email', $this->calendar_email)
                ->where('is_active', true)
                ->first();
        }

        return null;
    }

    public function getRouteKeyName()
    {
        return 'slug';
    }

    public function getPublicUrlAttribute()
    {
        return route('community.show', $this->slug);
    }

    public function getLogoUrlAttribute()
    {
        if ($this->logo) {
            return asset('storage/' . $this->logo);
        }

        return 'https://ui-avatars.com/api/?name=' . urlencode($this->name) . '&background=' . ltrim($this->color, '#') . '&color=fff&size=200';
    }

    public function scopePublic($query)
    {
        return $query->where('is_public', true)->where('is_active', true);
    }

    public function scopeForUser($query, $userId)
    {
        return $query->where('user_id', $userId)->where('is_active', true);
    }
}

File: synqs/lib/data/models/community.dart
import 'package:json_annotation/json_annotation.dart';
import 'user.dart';
import 'google_account.dart';
import 'calendar.dart';

part 'community.g.dart';

@JsonSerializable()
class Community {
  final int id;
  final String name;
  final String slug;
  final String? description;
  final String? logo;
  final String? website;
  @JsonKey(name: 'contact_email')
  final String? contactEmail;
  @JsonKey(name: 'calendar_email')
  final String? calendarEmail;
  @JsonKey(name: 'destination_calendar_id')
  final int? destinationCalendarId;
  @JsonKey(name: 'google_account_id')
  final int? googleAccountId;
  @JsonKey(name: 'social_links')
  final Map<String, dynamic>? socialLinks;
  final String? timezone;
  final String color;
  @JsonKey(name: 'is_public')
  final bool isPublic;
  @JsonKey(name: 'is_active')
  final bool isActive;
  @JsonKey(name: 'user_id')
  final int userId;
  final User? user;
  @JsonKey(name: 'google_account')
  final GoogleAccount? googleAccount;
  @JsonKey(name: 'destination_calendar')
  final Calendar? destinationCalendar;
  @JsonKey(name: 'events_count')
  final int? eventsCount;
  @JsonKey(name: 'calls_for_speakers_count')
  final int? callsForSpeakersCount;
  @JsonKey(name: 'logo_url')
  final String? logoUrl;
  @JsonKey(name: 'public_url')
  final String? publicUrl;
  @JsonKey(name: 'created_at')
  final DateTime? createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime? updatedAt;

  const Community({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    this.logo,
    this.website,
    this.contactEmail,
    this.calendarEmail,
    this.destinationCalendarId,
    this.googleAccountId,
    this.socialLinks,
    this.timezone,
    this.color = '#3B82F6',
    this.isPublic = true,
    this.isActive = true,
    required this.userId,
    this.user,
    this.googleAccount,
    this.destinationCalendar,
    this.eventsCount,
    this.callsForSpeakersCount,
    this.logoUrl,
    this.publicUrl,
    this.createdAt,
    this.updatedAt,
  });

  factory Community.fromJson(Map<String, dynamic> json) =>
      _$CommunityFromJson(json);

  Map<String, dynamic> toJson() => _$CommunityToJson(this);

  Community copyWith({
    int? id,
    String? name,
    String? slug,
    String? description,
    String? logo,
    String? website,
    String? contactEmail,
    String? calendarEmail,
    int? destinationCalendarId,
    int? googleAccountId,
    Map<String, dynamic>? socialLinks,
    String? timezone,
    String? color,
    bool? isPublic,
    bool? isActive,
    int? userId,
    User? user,
    GoogleAccount? googleAccount,
    Calendar? destinationCalendar,
    int? eventsCount,
    int? callsForSpeakersCount,
    String? logoUrl,
    String? publicUrl,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Community(
      id: id ?? this.id,
      name: name ?? this.name,
      slug: slug ?? this.slug,
      description: description ?? this.description,
      logo: logo ?? this.logo,
      website: website ?? this.website,
      contactEmail: contactEmail ?? this.contactEmail,
      calendarEmail: calendarEmail ?? this.calendarEmail,
      destinationCalendarId: destinationCalendarId ?? this.destinationCalendarId,
      googleAccountId: googleAccountId ?? this.googleAccountId,
      socialLinks: socialLinks ?? this.socialLinks,
      timezone: timezone ?? this.timezone,
      color: color ?? this.color,
      isPublic: isPublic ?? this.isPublic,
      isActive: isActive ?? this.isActive,
      userId: userId ?? this.userId,
      user: user ?? this.user,
      googleAccount: googleAccount ?? this.googleAccount,
      destinationCalendar: destinationCalendar ?? this.destinationCalendar,
      eventsCount: eventsCount ?? this.eventsCount,
      callsForSpeakersCount: callsForSpeakersCount ?? this.callsForSpeakersCount,
      logoUrl: logoUrl ?? this.logoUrl,
      publicUrl: publicUrl ?? this.publicUrl,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  bool get hasCalendarIntegration => calendarEmail != null || googleAccountId != null;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Community && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'Community(id: $id, name: $name, slug: $slug)';
  }
}